module/zfs/dmu.c

```diff
@@ -2138,6 +2140,145 @@ dmu_offset_next(objset_t *os, uint64_t object, boolean_t hole, uint64_t *off)
 	return (err);
 }
 
+int
+dmu_brt_readbps(objset_t *os, uint64_t object, uint64_t offset, uint64_t length,
+    dmu_tx_t *tx, blkptr_t **bpsp, size_t *nbpsp)
+{
+	dmu_buf_t **dbp, *dbuf;
+	dmu_buf_impl_t *db;
+	blkptr_t *bp, *bps;
+	int error, ii, numbufs;
+
+	error = dmu_buf_hold_array(os, object, offset, length, FALSE, FTAG,
+	    &numbufs, &dbp);
+	if (error != 0) {
+		if (error == ESRCH) {
+			error = SET_ERROR(ENXIO);
+		}
+		return (error);
+	}
+
+	bps = kmem_alloc(sizeof (blkptr_t) * numbufs, KM_SLEEP);
+
+	for (ii = 0; ii < numbufs; ii++) {
+		dbuf = dbp[ii];
+		db = (dmu_buf_impl_t *)dbuf;
+		bp = db->db_blkptr;
+
+		/*
+		 * If the block is not on the disk yet (so has no BP assigned),
+		 * there is not much we can do...
+		 */
+		if (!list_is_empty(&db->db_dirty_records)) {
+			dbuf_dirty_record_t *dr;
+
+			dr = list_head(&db->db_dirty_records);
+			if (dr->dt.dl.dr_brtwrite) {
+				/*
+				 * This is very special case where we clone a
+				 * file and in the same transaction group we
+				 * clone the clone.
+				 */
+				bp = &dr->dt.dl.dr_overridden_by;
+			} else {
+				/*
+				 * The block we want to clone was modified in
+				 * the same transaction group.
+				 */
+				error = SET_ERROR(EAGAIN);
+				goto out;
+			}
+		}
+		if (bp == NULL) {
+			/*
+			 * The block was created in this transaction group,
+			 * so it is too early to clone it.
+			 */
+			error = SET_ERROR(EAGAIN);
+			goto out;
+		}
+		if (dmu_buf_is_dirty(dbuf, tx)) {
+			error = SET_ERROR(EAGAIN);
+			goto out;
+		}
+		/*
+		 * Make sure we clone only data blocks.
+		 */
+		if (BP_IS_METADATA(bp)) {
+			error = SET_ERROR(EINVAL);
+			goto out;
+		}
+
+		bps[ii] = *bp;
+	}
+
+	*nbpsp = numbufs;
+	*bpsp = bps;
+out:
+	dmu_buf_rele_array(dbp, numbufs, FTAG);
+	if (error != 0) {
+		kmem_free(bps, sizeof (blkptr_t) * numbufs);
+	}
+
+	return (error);
+}
+
+void
+dmu_brt_addref(objset_t *os, uint64_t object, uint64_t offset, uint64_t length,
+    dmu_tx_t *tx, const blkptr_t *bps, size_t nbps)
+{
+	dmu_buf_t **dbp, *dbuf;
+	dmu_buf_impl_t *db;
+	struct dirty_leaf *dl;
+	dbuf_dirty_record_t *dr;
+	const blkptr_t *bp;
+	int numbufs;
+	uint_t ii;
+
+	VERIFY0(dmu_buf_hold_array(os, object, offset, length, FALSE, FTAG,
+	    &numbufs, &dbp));
+	ASSERT3U(nbps, ==, numbufs);
+
+	for (ii = 0; ii < numbufs; ii++) {
+		dbuf = dbp[ii];
+		db = (dmu_buf_impl_t *)dbuf;
+		bp = &bps[ii];
+
+		ASSERT0(db->db_level);
+		ASSERT(db->db_blkid != DMU_BONUS_BLKID);
+		ASSERT(BP_IS_HOLE(bp) || dbuf->db_size == BP_GET_LSIZE(bp));
+
+		dmu_buf_will_not_fill(dbuf, tx);
+
+		dr = list_head(&db->db_dirty_records);
+		ASSERT3U(dr->dr_txg, ==, tx->tx_txg);
+		dl = &dr->dt.dl;
+		dl->dr_overridden_by = *bp;
+		dl->dr_brtwrite = B_TRUE;
+
+		dl->dr_override_state = DR_OVERRIDDEN;
+		if (BP_IS_HOLE(bp)) {
+			dl->dr_overridden_by.blk_phys_birth = 0;
+			dl->dr_overridden_by.blk_birth = 0;
```

Comment:
Is this correct? This case is equivalent of hole being punched after being written, so as I understand it should get birth time of this TXG to replicate correctly.  Exception can be made only if it was hole and still remains a hole, in which case we may keep the pointer intact.

Reply:
Does that look better?
`--- a/module/zfs/dmu.c
+++ b/module/zfs/dmu.c
@@ -2272,12 +2272,11 @@ dmu_brt_addref(objset_t *os, uint64_t object, uint64_t offset, uint64_t length,
                dl->dr_override_state = DR_OVERRIDDEN;
                if (BP_IS_HOLE(bp)) {
                        dl->dr_overridden_by.blk_phys_birth = 0;
-                       dl->dr_overridden_by.blk_birth = 0;
                } else {
                        dl->dr_overridden_by.blk_phys_birth =
                            BP_PHYSICAL_BIRTH(bp);
-                       dl->dr_overridden_by.blk_birth = dr->dr_txg;
                }
+               dl->dr_overridden_by.blk_birth = dr->dr_txg;

                /*
                 * When data in embedded into BP there is no need to create`

Reply:
Probably, but I am not completely confident in this db_dirty_records area.  Additionally if both source and destination are holes, I guess this could be skipped completely to not increase metadata and replication traffic, but with the same caution.